server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/LinearFillOperator.java

```diff
@@ -0,0 +1,189 @@
+    return operatorContext;
+  }
+
+  @Override
+  public ListenableFuture<Void> isBlocked() {
+    return child.isBlocked();
+  }
+
+  @Override
+  public TsBlock next() {
+
+    // make sure we call child.next() at most once
+    if (cachedTsBlock.isEmpty()) {
+      canCallNext = false;
+      TsBlock nextBlock = child.next();
+      // child operator's calculation is not finished, so we just return null
+      if (nextBlock == null || nextBlock.isEmpty()) {
+        return nextBlock;
+      } else { // otherwise, we cache it
+        cachedTsBlock.add(nextBlock);
+      }
+    }
+
+    TsBlock block = cachedTsBlock.get(0);
+    long currentEndTime = block.getEndTime();
+    // use cached TsBlock to keep filling remaining column
+    while (currentFilledColumnIndex < outputColumnCount) {
+      // current valueColumn can't be filled using current information
+      if (fillArray[currentFilledColumnIndex].needPrepareForNext(
+          currentEndTime, block.getColumn(currentFilledColumnIndex))) {
+
+        if (canCallNext) { // if we can call child.next(), we call that and cache it in
+          // cachedTsBlock
+          canCallNext = false;
+          TsBlock nextBlock = child.next();
+          // child operator's calculation is not finished, so we just return null
+          if (nextBlock == null || nextBlock.isEmpty()) {
+            return nextBlock;
+          } else { // otherwise, we cache it
+            cachedTsBlock.add(nextBlock);
+          }
+        }
+
+        // next TsBlock has already been in the cachedTsBlock
+        while (nextTsBlockIndex[currentFilledColumnIndex] < cachedTsBlock.size()) {
+          TsBlock nextTsBlock = cachedTsBlock.get(nextTsBlockIndex[currentFilledColumnIndex]);
+          nextTsBlockIndex[currentFilledColumnIndex]++;
+          if (fillArray[currentFilledColumnIndex].prepareForNext(
+              currentEndTime,
+              nextTsBlock.getTimeColumn(),
+              nextTsBlock.getColumn(currentFilledColumnIndex))) {
+            cachedFilledValueColumns[currentFilledColumnIndex] =
+                fillArray[currentFilledColumnIndex].fill(
+                    block.getTimeColumn(), block.getColumn(currentFilledColumnIndex));
+            break;
+          }
+        }
+
+        // current column's filling is not finished using current owning TsBlock
+        if (cachedFilledValueColumns[currentFilledColumnIndex] == null) {
+          if (noMoreTsBlock) { // there is no more TsBlock, just use current owing TsBlock to fill
+            cachedFilledValueColumns[currentFilledColumnIndex] =
+                fillArray[currentFilledColumnIndex].fill(
+                    block.getTimeColumn(), block.getColumn(currentFilledColumnIndex));
+          } else { // next TsBlock is not ready, more TsBlocks is needed to do current column's
+            // fill, so current calculation is not finished, and we just return null
+            return null;
+          }
```

Comment:
From my perspective, it's easy to control time slice that you return null when you can't get enough information. But I still don't think It's a good way. The code is a bit confusing. I will support just getting next tsBlock until you can fill value using current information. Then the code will be very clear. You fill the value, if you can't, you get the following tsBlock.

Reply:
If so, it won't be consistent with our design philosophy. We should not let Operator to keep calling child.next, it should only call once because before you call next, you should also call isBlocked and hasNext which will also make the Operator code confusing, you should also dealing with blocked situation in each Operator.
Confusing code is not bad if it's correct and well-annotated.